File: Src/java/cql-to-elm/src/jvmTest/java/org/cqframework/cql/cql2elm/NamespaceTests.java
package org.cqframework.cql.cql2elm;

import static kotlinx.io.CoreKt.buffered;
import static kotlinx.io.JvmCoreKt.asSource;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.hamcrest.Matchers.notNullValue;

import java.io.IOException;
import kotlinx.io.Source;
import org.hl7.cql.model.NamespaceInfo;
import org.hl7.cql.model.NamespaceManager;
import org.hl7.cql_annotations.r1.CqlToElmError;
import org.hl7.elm.r1.IncludeDef;
import org.hl7.elm.r1.VersionedIdentifier;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

class NamespaceTests {
    private static ModelManager modelManager;
    private static LibraryManager libraryManager;
    private static NamespaceInfo defaultNamespaceInfo;
    private static NamespaceInfo coreNamespaceInfo;
    private static NamespaceInfo sharedNamespaceInfo;
    private static NamespaceInfo contentNamespaceInfo;

    static class NamespaceTestsLibrarySourceProvider implements LibrarySourceProvider {
        @Override
        public Source getLibrarySource(VersionedIdentifier libraryIdentifier) {
            String namespacePath = "NamespaceTests/";
            if (libraryIdentifier.getSystem() != null) {
                NamespaceInfo namespaceInfo =
                        libraryManager.getNamespaceManager().getNamespaceInfoFromUri(libraryIdentifier.getSystem());
                if (namespaceInfo != null && !namespaceInfo.getName().equals("Public")) {
                    namespacePath += namespaceInfo.getName() + "/";
                }
            }
            String libraryFileName = String.format(
                    "%s%s%s.cql",
                    namespacePath,
                    libraryIdentifier.getId(),
                    libraryIdentifier.getVersion() != null ? ("-" + libraryIdentifier.getVersion()) : "");
            var inputStream = org.cqframework.cql.cql2elm.NamespaceTests.class.getResourceAsStream(libraryFileName);
            if (inputStream == null) {
                return null;
            }
            return buffered(asSource(inputStream));
        }

        @Override
        public Source getLibraryContent(VersionedIdentifier libraryIdentifier, LibraryContentType type) {
            if (LibraryContentType.CQL == type) {
                return getLibrarySource(libraryIdentifier);
            }

            return null;
        }
    }

    @BeforeAll
    static void setup() {
        modelManager = new ModelManager();
        libraryManager = new LibraryManager(modelManager);
        libraryManager.getLibrarySourceLoader().registerProvider(new NamespaceTestsLibrarySourceProvider());
        defaultNamespaceInfo = new NamespaceInfo("Public", "http://cql.hl7.org/public");
        coreNamespaceInfo = new NamespaceInfo("Core", "http://cql.hl7.org/core");
        sharedNamespaceInfo = new NamespaceInfo("Shared", "http://cql.hl7.org/shared");
        contentNamespaceInfo = new NamespaceInfo("Content", "http://cql.hl7.org/content");
        libraryManager.getNamespaceManager().addNamespace(defaultNamespaceInfo);
        libraryManager.getNamespaceManager().addNamespace(coreNamespaceInfo);
        libraryManager.getNamespaceManager().addNamespace(sharedNamespaceInfo);
        libraryManager.getNamespaceManager().addNamespace(contentNamespaceInfo);
    }

    /* Unit test namespace functionality */

    @Test
    void namespacePath() {
        assertThat(
                NamespaceManager.Companion.getPath(defaultNamespaceInfo.getUri(), "Main"),
                is("http://cql.hl7.org/public/Main"));
    }

    @Test
    void namespaceNamePart() {
        assertThat(NamespaceManager.Companion.getNamePart("http://cql.hl7.org/public/Main"), is("Main"));
    }

    @Test
    void namespaceUriPart() {
        assertThat(
                NamespaceManager.Companion.getUriPart("http://cql.hl7.org/public/Main"),
                is("http://cql.hl7.org/public"));
    }

    /* Ensure base functionality with a defaulted namespace uri */

    @Test
    void libraryReferences() {
        CqlTranslator translator = null;
        try {
            translator = CqlTranslator.fromStream(
                    defaultNamespaceInfo,
                    NamespaceTests.class.getResourceAsStream("NamespaceTests/ReferencingLibrary.cql"),
                    libraryManager);
            assertThat(translator.getErrors().size(), is(0));
            assertThat(translator.toELM().getIdentifier().getSystem(), is(defaultNamespaceInfo.getUri()));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void invalidLibraryReferences() {
        CqlTranslator translator = null;
        try {
            translator = CqlTranslator.fromStream(
                    defaultNamespaceInfo,
                    NamespaceTests.class.getResourceAsStream("NamespaceTests/InvalidReferencingLibrary.cql"),
                    libraryManager);
            assertThat(translator.getErrors().size(), is(not(0)));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void invalidLibraryReference() {
        CqlTranslator translator = null;
        try {
            translator = CqlTranslator.fromStream(
                    defaultNamespaceInfo,
                    NamespaceTests.class.getResourceAsStream("NamespaceTests/InvalidLibraryReference.cql"),
                    libraryManager);
            assertThat(translator.getErrors().size(), is(not(0)));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void invalidBaseLibrary() {
        CqlTranslator translator = null;
        try {
            translator = CqlTranslator.fromStream(
                    defaultNamespaceInfo,
                    NamespaceTests.class.getResourceAsStream("NamespaceTests/ReferencingInvalidBaseLibrary.cql"),
                    libraryManager);
            assertThat(translator.getErrors().size(), is(1));
            assertThat(translator.getErrors().get(0), instanceOf(CqlCompilerException.class));
            assertThat(translator.getErrors().get(0).getLocator(), notNullValue());
            assertThat(translator.getErrors().get(0).getLocator().getLibrary(), notNullValue());
            assertThat(
                    translator.getErrors().get(0).getLocator().getLibrary().getSystem(),
                    is(defaultNamespaceInfo.getUri()));
            assertThat(translator.getErrors().get(0).getLocator().getLibrary().getId(), is("InvalidBaseLibrary"));

            assertThat(translator.toELM(), notNullValue());
            assertThat(translator.toELM().getAnnotation(), notNullValue());
            assertThat(translator.toELM().getAnnotation().size(), greaterThan(0));
            CqlToElmError invalidBaseLibraryError = null;
            for (Object o : translator.toELM().getAnnotation()) {
                if (o instanceof CqlToElmError) {
                    invalidBaseLibraryError = (CqlToElmError) o;
                    break;
                }
            }
            assertThat(invalidBaseLibraryError, notNullValue());
            assertThat(invalidBaseLibraryError.getLibrarySystem(), is(defaultNamespaceInfo.getUri()));
            assertThat(invalidBaseLibraryError.getLibraryId(), is("InvalidBaseLibrary"));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void main() {
        try {
            CqlTranslator translator = CqlTranslator.fromStream(
                    defaultNamespaceInfo,
                    NamespaceTests.class.getResourceAsStream("NamespaceTests/Main.cql"),
                    libraryManager);
            assertThat(translator.getErrors().size(), is(0));
            assertThat(translator.toELM().getIdentifier().getId(), is("Main"));
            assertThat(translator.toELM().getIdentifier().getSystem(), is(defaultNamespaceInfo.getUri()));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void referencingMain() {
        try {
            CqlTranslator translator = CqlTranslator.fromStream(
                    defaultNamespaceInfo,
                    NamespaceTests.class.getResourceAsStream("NamespaceTests/ReferencingMain.cql"),
                    libraryManager);
            assertThat(translator.getErrors().size(), is(0));
            assertThat(translator.toELM().getIdentifier().getId(), is("ReferencingMain"));
            assertThat(translator.toELM().getIdentifier().getSystem(), is(defaultNamespaceInfo.getUri()));
            assertThat(translator.toELM().getIncludes(), notNullValue());
            assertThat(translator.toELM().getIncludes().getDef(), notNullValue());
            assertThat(translator.toELM().getIncludes().getDef().size(), greaterThanOrEqualTo(1));
            IncludeDef i = translator.toELM().getIncludes().getDef().get(0);
            assertThat(i.getLocalIdentifier(), is("Main"));
            assertThat(i.getPath(), is("http://cql.hl7.org/public/Main"));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void coreMain() {
        try {
            CqlTranslator translator = CqlTranslator.fromStream(
                    coreNamespaceInfo,
                    NamespaceTests.class.getResourceAsStream("NamespaceTests/Core/Main.cql"),
                    libraryManager);
            assertThat(translator.getErrors().size(), is(0));
            assertThat(translator.toELM().getIdentifier().getId(), is("Main"));
            assertThat(translator.toELM().getIdentifier().getSystem(), is(coreNamespaceInfo.getUri()));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void coreReferencingMain() {
        try {
            CqlTranslator translator = CqlTranslator.fromStream(
                    coreNamespaceInfo,
                    NamespaceTests.class.getResourceAsStream("NamespaceTests/Core/ReferencingMain.cql"),
                    libraryManager);
            assertThat(translator.getErrors().size(), is(0));
            assertThat(translator.toELM().getIdentifier().getId(), is("ReferencingMain"));
            assertThat(translator.toELM().getIdentifier().getSystem(), is(coreNamespaceInfo.getUri()));
            assertThat(translator.toELM().getIncludes(), notNullValue());
            assertThat(translator.toELM().getIncludes().getDef(), notNullValue());
            assertThat(translator.toELM().getIncludes().getDef().size(), greaterThanOrEqualTo(1));
            IncludeDef i = translator.toELM().getIncludes().getDef().get(0);
            assertThat(i.getLocalIdentifier(), is("Main"));
            assertThat(i.getPath(), is("http://cql.hl7.org/core/Main"));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void sharedMain() {
        try {
            CqlTranslator translator = CqlTranslator.fromStream(
                    sharedNamespaceInfo,
                    NamespaceTests.class.getResourceAsStream("NamespaceTests/Shared/Main.cql"),
                    libraryManager);
            assertThat(translator.getErrors().size(), is(0));
            assertThat(translator.toELM().getIdentifier().getId(), is("Main"));
            assertThat(translator.toELM().getIdentifier().getSystem(), is(sharedNamespaceInfo.getUri()));

            assertThat(translator.toELM().getIncludes(), notNullValue());
            assertThat(translator.toELM().getIncludes().getDef(), notNullValue());
            assertThat(translator.toELM().getIncludes().getDef().size(), greaterThanOrEqualTo(1));
            IncludeDef i = translator.toELM().getIncludes().getDef().get(0);
            assertThat(i.getLocalIdentifier(), is("Main"));
            assertThat(i.getPath(), is("http://cql.hl7.org/core/Main"));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void sharedReferencingMain() {
        try {
            CqlTranslator translator = CqlTranslator.fromStream(
                    sharedNamespaceInfo,
                    NamespaceTests.class.getResourceAsStream("NamespaceTests/Shared/ReferencingMain.cql"),
                    libraryManager);
            assertThat(translator.getErrors().size(), is(0));
            assertThat(translator.toELM().getIdentifier().getId(), is("ReferencingMain"));
            assertThat(translator.toELM().getIdentifier().getSystem(), is(sharedNamespaceInfo.getUri()));
            assertThat(translator.toELM().getIncludes(), notNullValue());
            assertThat(translator.toELM().getIncludes().getDef(), notNullValue());
            assertThat(translator.toELM().getIncludes().getDef().size(), greaterThanOrEqualTo(1));
            IncludeDef i = translator.toELM().getIncludes().getDef().get(0);
            assertThat(i.getLocalIdentifier(), is("Main"));
            assertThat(i.getPath(), is("http://cql.hl7.org/shared/Main"));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void contentMain() {
        try {
            CqlTranslator translator = CqlTranslator.fromStream(
                    contentNamespaceInfo,
                    NamespaceTests.class.getResourceAsStream("NamespaceTests/Content/Main.cql"),
                    libraryManager);
            assertThat(translator.getErrors().size(), is(0));
            assertThat(translator.toELM().getIdentifier().getId(), is("Main"));
            assertThat(translator.toELM().getIdentifier().getSystem(), is(contentNamespaceInfo.getUri()));

            assertThat(translator.toELM().getIncludes(), notNullValue());
            assertThat(translator.toELM().getIncludes().getDef(), notNullValue());
            assertThat(translator.toELM().getIncludes().getDef().size(), greaterThanOrEqualTo(2));
            IncludeDef i = translator.toELM().getIncludes().getDef().get(0);
            assertThat(i.getLocalIdentifier(), is("CoreMain"));
            assertThat(i.getPath(), is("http://cql.hl7.org/core/Main"));

            i = translator.toELM().getIncludes().getDef().get(1);
            assertThat(i.getLocalIdentifier(), is("SharedMain"));
            assertThat(i.getPath(), is("http://cql.hl7.org/shared/Main"));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void contentReferencingMain() {
        try {
            CqlTranslator translator = CqlTranslator.fromStream(
                    contentNamespaceInfo,
                    NamespaceTests.class.getResourceAsStream("NamespaceTests/Content/ReferencingMain.cql"),
                    libraryManager);
            assertThat(translator.getErrors().size(), is(0));
            assertThat(translator.toELM().getIdentifier().getId(), is("ReferencingMain"));
            assertThat(translator.toELM().getIdentifier().getSystem(), is(contentNamespaceInfo.getUri()));
            assertThat(translator.toELM().getIncludes(), notNullValue());
            assertThat(translator.toELM().getIncludes().getDef(), notNullValue());
            assertThat(translator.toELM().getIncludes().getDef().size(), greaterThanOrEqualTo(3));

            IncludeDef i = translator.toELM().getIncludes().getDef().get(0);
            assertThat(i.getLocalIdentifier(), is("Main"));
            assertThat(i.getPath(), is("http://cql.hl7.org/shared/Main"));

            i = translator.toELM().getIncludes().getDef().get(1);
            assertThat(i.getLocalIdentifier(), is("ReferencingMain"));
            assertThat(i.getPath(), is("http://cql.hl7.org/core/ReferencingMain"));

            i = translator.toELM().getIncludes().getDef().get(2);
            assertThat(i.getLocalIdentifier(), is("SharedReferencingMain"));
            assertThat(i.getPath(), is("http://cql.hl7.org/shared/ReferencingMain"));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: cql-to-elm/src/jvmTest/kotlin/org/cqframework/cql/cql2elm/NamespaceTests.kt
package org.cqframework.cql.cql2elm

import java.io.IOException
import java.util.Locale
import kotlinx.io.Source
import kotlinx.io.asSource
import kotlinx.io.buffered
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers
import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.not
import org.hl7.cql.model.NamespaceInfo
import org.hl7.cql.model.NamespaceManager
import org.hl7.cql_annotations.r1.CqlToElmError
import org.hl7.elm.r1.Library
import org.hl7.elm.r1.VersionedIdentifier
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test

@Suppress("PrintStackTrace")
internal class NamespaceTests {
    internal class NamespaceTestsLibrarySourceProvider : LibrarySourceProvider {
        override fun getLibrarySource(libraryIdentifier: VersionedIdentifier): Source {
            var namespacePath = "NamespaceTests/"
            if (libraryIdentifier.system != null) {
                val namespaceInfo: NamespaceInfo? =
                    libraryManager!!
                        .namespaceManager
                        .getNamespaceInfoFromUri(libraryIdentifier.system!!)
                if (namespaceInfo != null && namespaceInfo.name != "Public") {
                    namespacePath += namespaceInfo.name + "/"
                }
            }
            val libraryFileName =
                String.format(
                    Locale.US,
                    "%s%s%s.cql",
                    namespacePath,
                    libraryIdentifier.id,
                    if (libraryIdentifier.version != null) ("-" + libraryIdentifier.version) else "",
                )
            return NamespaceTests::class
                .java
                .getResourceAsStream(libraryFileName)!!
                .asSource()
                .buffered()
        }

        override fun getLibraryContent(
            libraryIdentifier: VersionedIdentifier,
            type: LibraryContentType,
        ): Source? {
            if (LibraryContentType.CQL == type) {
                return getLibrarySource(libraryIdentifier)
            }

            return null
        }
    }

    /* Unit test namespace functionality */
    @Test
    fun namespacePath() {
        assertThat(
            NamespaceManager.getPath(defaultNamespaceInfo!!.uri, "Main"),
            `is`("http://cql.hl7.org/public/Main"),
        )
    }

    @Test
    fun namespaceNamePart() {
        assertThat<String?>(
            NamespaceManager.getNamePart("http://cql.hl7.org/public/Main"),
            `is`<String?>("Main"),
        )
    }

    @Test
    fun namespaceUriPart() {
        assertThat<String?>(
            NamespaceManager.getUriPart("http://cql.hl7.org/public/Main"),
            `is`<String?>("http://cql.hl7.org/public"),
        )
    }

    /* Ensure base functionality with a defaulted namespace uri */
    @Test
    fun libraryReferences() {
        var translator: CqlTranslator?
        try {
            translator =
                CqlTranslator.fromSource(
                    defaultNamespaceInfo,
                    NamespaceTests::class
                        .java
                        .getResourceAsStream("NamespaceTests/ReferencingLibrary.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )
            assertThat(translator.errors.size, `is`(0))
            assertThat(translator.toELM()!!.identifier!!.system, `is`(defaultNamespaceInfo!!.uri))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun invalidLibraryReferences() {
        var translator: CqlTranslator?
        try {
            translator =
                CqlTranslator.fromSource(
                    defaultNamespaceInfo,
                    NamespaceTests::class
                        .java
                        .getResourceAsStream("NamespaceTests/InvalidReferencingLibrary.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )
            assertThat(translator.errors.size, Matchers.`is`(not(0)))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun invalidLibraryReference() {
        var translator: CqlTranslator?
        try {
            translator =
                CqlTranslator.fromSource(
                    defaultNamespaceInfo,
                    NamespaceTests::class
                        .java
                        .getResourceAsStream("NamespaceTests/InvalidLibraryReference.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )
            assertThat(translator.errors.size, Matchers.`is`(not(0)))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun invalidBaseLibrary() {
        var translator: CqlTranslator?
        try {
            translator =
                CqlTranslator.fromSource(
                    defaultNamespaceInfo,
                    NamespaceTests::class
                        .java
                        .getResourceAsStream("NamespaceTests/ReferencingInvalidBaseLibrary.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )
            assertThat(translator.errors.size, `is`(1))
            assertThat(translator.errors[0], Matchers.instanceOf(CqlCompilerException::class.java))
            assertThat(translator.errors[0].locator, Matchers.notNullValue())
            assertThat(translator.errors[0].locator!!.library, Matchers.notNullValue())
            assertThat(
                translator.errors[0].locator!!.library!!.system,
                `is`(defaultNamespaceInfo!!.uri),
            )
            assertThat(translator.errors[0].locator!!.library!!.id, `is`("InvalidBaseLibrary"))

            assertThat<Library?>(translator.toELM(), Matchers.notNullValue())
            assertThat<Any?>(translator.toELM()!!.annotation, Matchers.notNullValue())
            assertThat(translator.toELM()!!.annotation.size, Matchers.greaterThan(0))
            var invalidBaseLibraryError: CqlToElmError? = null
            for (o in translator.toELM()!!.annotation) {
                if (o is CqlToElmError) {
                    invalidBaseLibraryError = o
                    break
                }
            }
            assertThat<CqlToElmError?>(invalidBaseLibraryError, Matchers.notNullValue())
            assertThat(invalidBaseLibraryError!!.librarySystem, `is`(defaultNamespaceInfo!!.uri))
            assertThat<String?>(
                invalidBaseLibraryError.libraryId,
                `is`<String?>("InvalidBaseLibrary"),
            )
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun main() {
        try {
            val translator =
                CqlTranslator.fromSource(
                    defaultNamespaceInfo,
                    NamespaceTests::class
                        .java
                        .getResourceAsStream("NamespaceTests/Main.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )
            assertThat(translator.errors.size, `is`(0))
            assertThat<String?>(translator.toELM()!!.identifier!!.id, `is`<String?>("Main"))
            assertThat(translator.toELM()!!.identifier!!.system, `is`(defaultNamespaceInfo!!.uri))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun referencingMain() {
        try {
            val translator =
                CqlTranslator.fromSource(
                    defaultNamespaceInfo,
                    NamespaceTests::class
                        .java
                        .getResourceAsStream("NamespaceTests/ReferencingMain.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )
            assertThat(translator.errors.size, `is`(0))
            assertThat<String?>(
                translator.toELM()!!.identifier!!.id,
                `is`<String?>("ReferencingMain"),
            )
            assertThat(translator.toELM()!!.identifier!!.system, `is`(defaultNamespaceInfo!!.uri))
            assertThat<Library.Includes?>(translator.toELM()!!.includes, Matchers.notNullValue())
            assertThat<Any?>(translator.toELM()!!.includes!!.def, Matchers.notNullValue())
            assertThat(translator.toELM()!!.includes!!.def.size, Matchers.greaterThanOrEqualTo(1))
            val i = translator.toELM()!!.includes!!.def[0]
            assertThat<String?>(i.localIdentifier, `is`<String?>("Main"))
            assertThat<String?>(i.path, `is`<String?>("http://cql.hl7.org/public/Main"))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun coreMain() {
        try {
            val translator =
                CqlTranslator.fromSource(
                    coreNamespaceInfo,
                    NamespaceTests::class
                        .java
                        .getResourceAsStream("NamespaceTests/Core/Main.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )
            assertThat(translator.errors.size, `is`(0))
            assertThat<String?>(translator.toELM()!!.identifier!!.id, `is`<String?>("Main"))
            assertThat(translator.toELM()!!.identifier!!.system, `is`(coreNamespaceInfo!!.uri))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun coreReferencingMain() {
        try {
            val translator =
                CqlTranslator.fromSource(
                    coreNamespaceInfo,
                    NamespaceTests::class
                        .java
                        .getResourceAsStream("NamespaceTests/Core/ReferencingMain.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )
            assertThat(translator.errors.size, `is`(0))
            assertThat<String?>(
                translator.toELM()!!.identifier!!.id,
                `is`<String?>("ReferencingMain"),
            )
            assertThat(translator.toELM()!!.identifier!!.system, `is`(coreNamespaceInfo!!.uri))
            assertThat<Library.Includes?>(translator.toELM()!!.includes, Matchers.notNullValue())
            assertThat<Any?>(translator.toELM()!!.includes!!.def, Matchers.notNullValue())
            assertThat(translator.toELM()!!.includes!!.def.size, Matchers.greaterThanOrEqualTo(1))
            val i = translator.toELM()!!.includes!!.def[0]
            assertThat<String?>(i.localIdentifier, `is`<String?>("Main"))
            assertThat<String?>(i.path, `is`<String?>("http://cql.hl7.org/core/Main"))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun sharedMain() {
        try {
            val translator =
                CqlTranslator.fromSource(
                    sharedNamespaceInfo,
                    NamespaceTests::class
                        .java
                        .getResourceAsStream("NamespaceTests/Shared/Main.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )
            assertThat(translator.errors.size, `is`(0))
            assertThat<String?>(translator.toELM()!!.identifier!!.id, `is`<String?>("Main"))
            assertThat(translator.toELM()!!.identifier!!.system, `is`(sharedNamespaceInfo!!.uri))

            assertThat<Library.Includes?>(translator.toELM()!!.includes, Matchers.notNullValue())
            assertThat<Any?>(translator.toELM()!!.includes!!.def, Matchers.notNullValue())
            assertThat(translator.toELM()!!.includes!!.def.size, Matchers.greaterThanOrEqualTo(1))
            val i = translator.toELM()!!.includes!!.def[0]
            assertThat<String?>(i.localIdentifier, `is`<String?>("Main"))
            assertThat<String?>(i.path, `is`<String?>("http://cql.hl7.org/core/Main"))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun sharedReferencingMain() {
        try {
            val translator =
                CqlTranslator.fromSource(
                    sharedNamespaceInfo,
                    NamespaceTests::class
                        .java
                        .getResourceAsStream("NamespaceTests/Shared/ReferencingMain.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )
            assertThat(translator.errors.size, `is`(0))
            assertThat<String?>(
                translator.toELM()!!.identifier!!.id,
                `is`<String?>("ReferencingMain"),
            )
            assertThat(translator.toELM()!!.identifier!!.system, `is`(sharedNamespaceInfo!!.uri))
            assertThat<Library.Includes?>(translator.toELM()!!.includes, Matchers.notNullValue())
            assertThat<Any?>(translator.toELM()!!.includes!!.def, Matchers.notNullValue())
            assertThat(translator.toELM()!!.includes!!.def.size, Matchers.greaterThanOrEqualTo(1))
            val i = translator.toELM()!!.includes!!.def[0]
            assertThat<String?>(i.localIdentifier, `is`<String?>("Main"))
            assertThat<String?>(i.path, `is`<String?>("http://cql.hl7.org/shared/Main"))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun contentMain() {
        try {
            val translator =
                CqlTranslator.fromSource(
                    contentNamespaceInfo,
                    NamespaceTests::class
                        .java
                        .getResourceAsStream("NamespaceTests/Content/Main.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )
            assertThat(translator.errors.size, `is`(0))
            assertThat<String?>(translator.toELM()!!.identifier!!.id, `is`<String?>("Main"))
            assertThat(translator.toELM()!!.identifier!!.system, `is`(contentNamespaceInfo!!.uri))

            assertThat<Library.Includes?>(translator.toELM()!!.includes, Matchers.notNullValue())
            assertThat<Any?>(translator.toELM()!!.includes!!.def, Matchers.notNullValue())
            assertThat(translator.toELM()!!.includes!!.def.size, Matchers.greaterThanOrEqualTo(2))
            var i = translator.toELM()!!.includes!!.def[0]
            assertThat<String?>(i.localIdentifier, `is`<String?>("CoreMain"))
            assertThat<String?>(i.path, `is`<String?>("http://cql.hl7.org/core/Main"))

            i = translator.toELM()!!.includes!!.def[1]
            assertThat<String?>(i.localIdentifier, `is`<String?>("SharedMain"))
            assertThat<String?>(i.path, `is`<String?>("http://cql.hl7.org/shared/Main"))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun contentReferencingMain() {
        try {
            val translator =
                CqlTranslator.fromSource(
                    contentNamespaceInfo,
                    NamespaceTests::class
                        .java
                        .getResourceAsStream("NamespaceTests/Content/ReferencingMain.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )
            assertThat(translator.errors.size, `is`(0))
            assertThat<String?>(
                translator.toELM()!!.identifier!!.id,
                `is`<String?>("ReferencingMain"),
            )
            assertThat(translator.toELM()!!.identifier!!.system, `is`(contentNamespaceInfo!!.uri))
            assertThat<Library.Includes?>(translator.toELM()!!.includes, Matchers.notNullValue())
            assertThat<Any?>(translator.toELM()!!.includes!!.def, Matchers.notNullValue())
            assertThat(translator.toELM()!!.includes!!.def.size, Matchers.greaterThanOrEqualTo(3))

            var i = translator.toELM()!!.includes!!.def[0]
            assertThat<String?>(i.localIdentifier, `is`<String?>("Main"))
            assertThat<String?>(i.path, `is`<String?>("http://cql.hl7.org/shared/Main"))

            i = translator.toELM()!!.includes!!.def[1]
            assertThat<String?>(i.localIdentifier, `is`<String?>("ReferencingMain"))
            assertThat<String?>(i.path, `is`<String?>("http://cql.hl7.org/core/ReferencingMain"))

            i = translator.toELM()!!.includes!!.def[2]
            assertThat<String?>(i.localIdentifier, `is`<String?>("SharedReferencingMain"))
            assertThat<String?>(i.path, `is`<String?>("http://cql.hl7.org/shared/ReferencingMain"))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    companion object {
        private var modelManager: ModelManager? = null
        private var libraryManager: LibraryManager? = null
        private var defaultNamespaceInfo: NamespaceInfo? = null
        private var coreNamespaceInfo: NamespaceInfo? = null
        private var sharedNamespaceInfo: NamespaceInfo? = null
        private var contentNamespaceInfo: NamespaceInfo? = null

        @JvmStatic
        @BeforeAll
        fun setup() {
            modelManager = ModelManager()
            libraryManager = LibraryManager(modelManager!!)
            libraryManager!!
                .librarySourceLoader
                .registerProvider(NamespaceTestsLibrarySourceProvider())
            defaultNamespaceInfo = NamespaceInfo("Public", "http://cql.hl7.org/public")
            coreNamespaceInfo = NamespaceInfo("Core", "http://cql.hl7.org/core")
            sharedNamespaceInfo = NamespaceInfo("Shared", "http://cql.hl7.org/shared")
            contentNamespaceInfo = NamespaceInfo("Content", "http://cql.hl7.org/content")
            libraryManager!!.namespaceManager.addNamespace(defaultNamespaceInfo!!)
            libraryManager!!.namespaceManager.addNamespace(coreNamespaceInfo!!)
            libraryManager!!.namespaceManager.addNamespace(sharedNamespaceInfo!!)
            libraryManager!!.namespaceManager.addNamespace(contentNamespaceInfo!!)
        }
    }
}
